text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+*Note: this is different from `mem::discriminant`.  For example,*
+```rust
+#![allow(enum_intrinsics_non_enums)]
+enum MyEnum { A, B }
+let a = Box::new(MyEnum::A);
+let b = Box::new(MyEnum::B);
+assert_eq!(std::mem::discriminant(&a), std::mem::discriminant(&b));
+assert_ne!(a.enum#discriminant, b.enum#discriminant);
+```
+<!-- https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=76293c2f83cb7719c22c15bcebeaeb13 -->
+
+For this, a generic parameter is never considered to be an `enum`,
+although a generic enum where some of the generic parameters to the
+enum constructor are not yet known is fine.
+
+It's an error if, despite deref'ing, the LHS is still not an `enum`.
+
+If the enum has `repr(uN)` or `repr(iM)`, the `.enum#discriminant` expression
+returns a value of type `uN` or `iM` respectively.
+
+If the enum does not specify an integer `repr`, then it returns `isize`.
+
+*Note: `isize` is rarely the desired type for discriminants, and indeed custom
+discriminants on types with fields are disallowed without explicit `repr` types.
+Returning `isize` is fine here, though, thanks to privacy because the code
+inside the module can be updated should it change to specify a specific type.*
+
+## Semantics
+
+When the LHS of a discriminant expression is a *place*, that place is read but not consumed.
+
+*Note: this can be thought of as if it read a field of `Copy` type from the LHS.*
+
+This lowers to [`Rvalue::Discriminant`][MIR discr] in MIR.
+
+[MIR discr]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html#variant.Discriminant
+
+As this expression is an *r-value*, not a *place*, `&foo.enum#discriminant` returns
+a reference to a temporary, aka is the same as `&{foo.enum#discriminant}`.
+It does *not* return a reference to the memory in which the discriminant is
+stored -- not even for types that do store the discriminant directly.
+
+This expression is allowed in `const` contexts, but is not promotable.
+
+*Note: the behaviour of this expression is independent of whether the type gets
+layout-optimized.  For example, the following holds even if `x` is `2_i8` in memory.*
+```rust
+enum MyOption<T> { MyNone, MySome(T) }
+let x = MyOption::<std::cmp::Ordering>::MyNone;
+assert_eq!(x.enum#discriminant, 0_isize);
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This isn't strictly necessary, we could continue to get along just fine without it.
+
+- For the FFI cases the layout guarantees mean it's already possible to write a
+  sound and reliable function that reads the discriminant.
+- For cases without `repr(int)`, custom discriminants aren't even allowed,
+  so those discriminants much not be all that important.
+- It's always possible to write a `match` in safe code that optimizes away
+  and produces exactly the same thing that this new expression would.
+- A pseudo-field with `#` in the name looks kinda weird.
+- There might be a nicer way to do this in the future.
+
+# Rationale and alternatives
```

Comment:
Another related alternative is creating a new derivable trait that exposes the discrimenant. Which allows the user to opt in to exposing the discrimenant, but doesn't have a way to make it accessible only within the crate.

Reply:
#3604 is irrelevant to this RFC besides field-less (C-style) enum. 

If extending #3604 to enum with fields, the `Into` implementation would be often useless (e.g. `Option<T>::into()` will return `if self.is_none() { 0 } else { 1 }`), and the `TryFrom` implementation will not work at all.

OTOH, accepting this RFC won't accomplish what is proposed in #3604 either, e.g. those enums can't be passed into arguments expecting `impl Into<u32>`.

----

> Another related alternative is creating a new derivable trait that exposes the discrimenant. Which allows the user to opt in to exposing the discrimenant, 

Given that `std::mem::Discriminant<E>` is already applicable to all types I don't think you need a derivative. Perhaps a marker trait is sufficient.

```rust
/* module std::marker */

/// An assertion that the enum has a stable `#[repr]` 
/// and every variant's discriminant value will never change.
///
/// In particular, enums with implicit discriminant values must not reorder their variants.
// (not sure if the trait needs to be unsafe or not)
pub trait StableDiscriminant {}

/* module std::mem */

impl<E: StableDiscriminant> Discriminant<E> {
    pub fn value(&self) -> <E as DiscriminantKind>::Discriminant {
        self.0
    }
}

/* user code */

enum Enum { ... }

impl std::marker::StableDiscriminant for Enum {}
```

**However** this will not solve what the RFC's motivation which is `#[derive(PartialOrd)]`, where implementing PartialOrd does not mean the freedom to reorder the enum variants should be taken away (one could change from derive to manual implementation to keep the PartialOrd stable while actually changing the discriminant value).

So the whole issue is still above whether the discriminant value of a regular enum should be considered part of the stable public interface. If not then the whole issue of

> but doesn't have a way to make it accessible only within the crate.

can just goes away.